projects/dsp-ui/src/lib/search/fulltext-search/fulltext-search.component.spec.ts

```diff
@@ -0,0 +1,399 @@
+        fulltextSearchComponentDe = hostCompDe.query(By.directive(FulltextSearchComponent));
+    });
+
+    // Mock localStorage
+    beforeEach(() => {
+        let store = {};
+
+        spyOn(localStorage, 'getItem').and.callFake((key: string): string => {
+            return store[key] || null;
+        });
+        spyOn(localStorage, 'removeItem').and.callFake((key: string): void => {
+            delete store[key];
+        });
+        spyOn(localStorage, 'setItem').and.callFake((key: string, value: string): string => {
+            return store[key] = value as string;
+        });
+        spyOn(localStorage, 'clear').and.callFake(() => {
+            store = {};
+        });
+    });
+
+    it('should create', () => {
+        expect(testHostComponent).toBeTruthy();
+        expect(testHostComponent.fulltextSearch).toBeTruthy();
+    });
+
+    it('should get projects on init', () => {
+        const projSpy = TestBed.inject(DspApiConnectionToken);
+
+        expect(testHostComponent.fulltextSearch.projects).toBeDefined();
+        expect(testHostComponent.fulltextSearch.projects.length).toEqual(8);
+        expect(testHostComponent.fulltextSearch.projectfilter).toEqual(true);
+
+        // TO FIX: expect(projSpy.admin.projectsEndpoint.getProjects).toHaveBeenCalledTimes(1);
+
+    });
+
+    it('should do a search', () => {
+        const prevSearchArray: PrevSearchItem[] = [
+            { projectIri: 'http://rdfh.ch/projects/0803', projectLabel: 'incunabula', query: 'one thing' }
+        ];
+
+        localStorage.setItem('prevSearch', JSON.stringify(prevSearchArray));
+        expect(localStorage.getItem('prevSearch')).toBeDefined();
+
+        const searchBtn = fulltextSearchComponentDe.query(By.css('button.kui-fulltext-search-button'));
+        expect(searchBtn).toBeDefined();
+        const searchInput = fulltextSearchComponentDe.query(By.css('input.kui-fulltext-search-input')).nativeElement;
+        expect(searchInput).toBeDefined();
+
+
+        searchInput.value = 'new thing';
+        searchInput.dispatchEvent(new Event('input'));
+
+        // click on the search button and trigger the method doSearch()
+        searchBtn.triggerEventHandler('click', null);
+        testHostFixture.detectChanges();
+
+        // check the local storage state
+        expect(testHostComponent.fulltextSearch.searchQuery).toEqual('new thing');
+        const newPrevSearchArray: PrevSearchItem[] = [
+            { projectIri: 'http://rdfh.ch/projects/0803', projectLabel: 'incunabula', query: 'one thing' },
+            { query: 'new thing' }
+        ];
+        expect(localStorage.getItem('prevSearch')).toEqual(JSON.stringify(newPrevSearchArray));
+
+        // check the call of router.navigate with correct arguments
+        expect(mockRouter.navigate).toHaveBeenCalledWith(['/search/fulltext/new thing']);
+
+    });
+
+    // todo: fix error: Cannot read property 'detach' of undefined CDK OVERLAY
```

Comment:
@kilchenmann @tobiasschweizer I get the following error: Cannot read property 'detach' of undefined. I suppose I have to define Overlay somewhere but I looked everywhere and tried several tutorials unsuccessfully.... Any idea?

Reply:
When I run the test "'should do a search'", I get the following errors:

![Screenshot 2020-06-23 at 12 52 59](https://user-images.githubusercontent.com/6000023/85395498-887b1400-b550-11ea-96fc-b4f5a569b750.png)

I am not sure where the zone errors (XMLHTTPRequest) come from. The router or some other dependency that should be mocked? 

I think the `detach` problem comes from an uninitialised `this.overlayRef`. In this case, it is `undefined`.

Reply:
yes this is my thought too but I could not find the proper way to initiate OverlayRef...

Reply:
maybe you could mock it like the router?

Reply:
@tobiasschweizer I have tried multiple ideas and tutorials, nothing seems to work

Reply:
Couldn't you just make an object with the methods you need and pass it to the `providers`? Otherwise we can look at this together.

Reply:
OverlayRef is part of the [CDK Overlay](https://material.angular.io/cdk/overlay/overview) from Angular Material. It is for me a box magic box and I have no idea how I should mock it or make an object of it with methods...